bot/code_review_bot/stats.py

```diff
@@ -3,51 +3,108 @@
+        self.client = InfluxDBClient(
+            conf["host"],
+            conf["port"],
+            conf["username"],
+            conf["password"],
+            conf["database"],
+            ssl=conf.get("ssl", False),
+            verify_ssl=conf.get("ssl", False),
+        )
+        assert self.client.ping()
+        logger.info(
+            "InfluxDb reporting enabled", database=conf["database"], host=conf["host"]
         )
-        self.api.constant_tags = ["code-review", "env:{}".format(settings.app_channel)]
-        self.api.start(flush_in_thread=True)
-        assert not self.api._disabled
+
+    def add_metric(self, name, value=1, tags={}):
+        """
+        Store a metric in memory, using InfluxDb point format
+        """
+        tags.update({"app": "code-review-bot", "channel": settings.app_channel})
+        self.metrics.append(
+            {
+                "measurement": f"code-review.{name}",
+                "tags": tags,
+                "time": datetime.utcnow().isoformat(),
+                "fields": {"value": value},
+            }
+        )
+
+    def flush(self):
+        """
+        Publish all metrics in memory to influxdb
+        """
+        if self.client is None:
+            logger.warning(
+                "InfluxDb client not connected: metrics will not be reported"
+            )
+            return
+        if not self.metrics:
+            return
+
+        logger.info("Flushing stats metrics", nb=len(self.metrics))
+        # TODO: add a retry ?
+        self.client.write_points(self.metrics)
+        self.metrics = []
 
     def report_task(self, task, issues):
         """
         Aggregate statistics about issues from a remote analysis task
         """
         assert isinstance(task, AnalysisTask)
+        tags = {"task": task.name}
 
         # Report all issues found
-        self.api.increment("issues.{}".format(task.name), len(issues))
+        self.add_metric("issues", len(issues), tags)
 
         # Report publishable issues
-        self.api.increment(
-            "issues.{}.publishable".format(task.name),
-            sum(i.is_publishable() for i in issues),
+        self.add_metric(
+            "issues.publishable", sum(i.is_publishable() for i in issues), tags
         )
 
         # Report total paths
-        self.api.increment(
-            "issues.{}.paths".format(task.name), len({i.path for i in issues})
-        )
+        self.add_metric("issues.paths", len({i.path for i in issues}), tags)
 
         # Report cleaned paths
-        self.api.increment(
-            "issues.{}.cleaned_paths".format(task.name), len(task.cleaned_paths)
-        )
+        self.add_metric("issues.cleaned_paths", len(task.cleaned_paths), tags)
+
+    @contextmanager
+    def timer(self, name):
+        """
+        A context manager tracking the contained code's runtime
+        """
+        start = time.time()
```

Comment:
Maybe better to use https://docs.python.org/3/library/time.html#time.perf_counter ?

Reply:
:+1:

Reply:
TIL